### PR TITLE
Creation of teamMembership collection. This tracks users memberships in teams

### PR DESCRIPTION
**Pull request by oriyomibadmus**

### CODE REVIEW

File: src/api/team-membership/content-types/team-membership/schema.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      "relation": "oneToOne",
+      "target": "plugin::users-permissions.user"
+    },
+    "joinDate": {
+      "type": "datetime"
+    },
+    "leaveDate": {
+      "type": "datetime"
+    },
+    "project": {
```
Comment (chungthuang) [inline src/api/team-membership/content-types/team-membership/schema.json:26]: I think this should be a required field.

File: src/api/team-membership/content-types/team-membership/schema.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    "pluralName": "team-memberships",
+    "displayName": "TeamMembership",
+    "description": ""
+  },
+  "options": {
+    "draftAndPublish": false
+  },
+  "pluginOptions": {},
+  "attributes": {
+    "user": {
```
Comment (chungthuang) [inline src/api/team-membership/content-types/team-membership/schema.json:15]: I think this should be a required field.

File: src/index.js [hunk 3]
```diff
@@ -16,5 +16,64 @@ module.exports = {
...
+      models: ["plugin::users-permissions.user"],
+      // afterCreate lifecycle to create a new team-membership entry when a new user is created
+      async afterCreate({result}){
+        const {projects} = result
+
+        for (const project of projects) {
+          await strapi.service("api::team-membership.team-membership").create({
+            data: { 
+              user: [result.id],
+              project: [project.id],
```
Comment (chungthuang) [inline src/index.js:31]: What if the user hasn't join a project? We shouldn't create an entry in team-membership if the user hasn't join a project,

File: src/index.js [hunk 4]
```diff
@@ -16,5 +16,64 @@ module.exports = {
...
+              joinDate: new Date()
+            }
+          })
+        }
+      },
+      
+      async beforeUpdate({params}) {
+        const joinProject = params.data.projects.connect;
```
Comment (chungthuang) [inline src/index.js:39]: What is `projects.connect`?

File: src/api/team-membership/content-types/team-membership/schema.json (hunk 1 above)
Comment (oriyomibadmus) [reply src/api/team-membership/content-types/team-membership/schema.json:26]: Okay, I also think I will have to make the default value "null" at first. This is cause the team-membership record for an active user still in a team won't have a `leaveDate` just yet

File: src/index.js (hunk 3 above)
Comment (oriyomibadmus) [reply src/index.js:31]: The` afterCreate` lifecycle triggers a `For loop` which loops over a list of projects the new `user` is a part of. If the project list is empty, the `strapi.service(...).create(...)` will not create a new entry in the `team-membership`

File: src/index.js (hunk 4 above)
Comment (oriyomibadmus) [reply src/index.js:39]: The `beforeUpdate` lifecycle holds the information about the new update that is about to be done on the user collection in the `params.data` object. This object also contains the projects object (`params.data.projects`). In the `parama.data.projects` objects, there are 2 arrays of objects [disconnect ] and [connect]:

`params.data.projects.disconnect` = List of projects removed by the user
`params.data.projects.connect` = List of projects added by the users

File: src/api/team-membership/content-types/team-membership/schema.json (hunk 2 above)
Comment (oriyomibadmus) [reply src/api/team-membership/content-types/team-membership/schema.json:15]: Okay, i will do that now

File: src/api/team-membership/content-types/team-membership/schema.json (hunk 1 above)
Comment (oriyomibadmus) [reply src/api/team-membership/content-types/team-membership/schema.json:26]: Which, you mean the `leaveDate` or `project` field ?

Comment (chungthuang) [reply src/api/team-membership/content-types/team-membership/schema.json:26]: The project field.

**Review by chungthuang — APPROVED** [review]

File: src/api/team-membership/content-types/team-membership/schema.json (hunk 1 above)
Comment (oriyomibadmus) [reply src/api/team-membership/content-types/team-membership/schema.json:27]: Done now

File: src/api/team-membership/content-types/team-membership/schema.json (hunk 2 above)
Comment (oriyomibadmus) [reply src/api/team-membership/content-types/team-membership/schema.json:15]: Done now

**Pull request merged**